frontend/catalyst/pennylane_extensions.py

```diff
@@ -166,20 +166,20 @@ def __call__(self, *args, **kwargs):
             # Allow QFunc to still be used by itself for internal testing.
             device = self.device
 
-        with EvaluationContext(EvaluationMode.QUANTUM_COMPILATION):
-            jaxpr, shape = trace_quantum_function(self.func, device, args, kwargs, qnode)
-
-        retval_tree = tree_structure(shape)
-
-        def _eval_jaxpr(*args):
-            return eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
-
-        args_data, _ = tree_flatten(args)
-
-        wrapped = wrap_init(_eval_jaxpr)
-        retval = func_p.bind(wrapped, *args_data, fn=self)
+        def _eval_quantum(*args):
+            closed_jaxpr, out_type, out_tree = trace_quantum_function(
+                self.func, device, args, kwargs, qnode
+            )
+            args_expanded = get_implicit_and_explicit_flat_args(None, *args)
+            res_expanded = eval_jaxpr(closed_jaxpr.jaxpr, closed_jaxpr.consts, *args_expanded)
+            _, out_keep = unzip2(out_type)
+            res_flat = [r for r, k in zip(res_expanded, out_keep) if k]
+            return tree_unflatten(out_tree, res_flat)
 
-        return tree_unflatten(retval_tree, retval)
+        wffa, _, _, out_tree_promise = deduce_avals(_eval_quantum, args, {})
```

Comment:
wffa means wrapped-flattened-annotated-function, I agree that it is overly cryptic :)  I plan to improve this whole `deduce_aval` part in the subsequent PR, because it is a rather fundamental part, but feel free to suggest a better name. Maybe `flattened_fun` because flatness is the most important property?

Reply:
Done as suggested